syntax = "proto3";

package yh_msg;

// 共用区
// 响应状态
message Status {
    int64 number = 1;
    int32 code = 2;
    string msg = 3;
}

// 标签
message Tag {
    int64 id = 1; // 标签ID(貌似)
    string text = 3;
    string color = 4;
}

message Msg {
    string msg_id = 1; // 消息ID
    Sender sender = 2;
    string direction = 3; // 消息位置,左边/右边
    int32 content_type = 4;
    Content content = 5;
    int64 send_time = 6; // 时间戳(毫秒)
    Cmd cmd = 7; // 指令
    int64 msg_delete_time = 8; // 消息撤回时间
    string quote_msg_id = 9; // 引用消息ID
    int64 msg_seq = 10;
    int64 edit_time = 12; // 最后编辑时间
        
    message Cmd {
        string name = 2; // 指令名
        int32 type = 4; // 指令类型
    }
    // 消息
    message Content {
        string text = 1; // 消息内容
        string buttons = 2; // 按钮
        string image_url = 3;
        string file_name = 4;
        string file_url = 5;
        string form = 7; // 表单消息
        string quote_msg_text = 8; // 引用消息文字
        string sticker_url = 9; // 表情URL
        string post_id = 10; // 文章ID
        string post_title = 11; // 文章标题
        string post_content = 12; // 文章内容
        string post_content_type = 13; // 文章类型
        string expression_id = 15; // 个人表情ID(不知道为啥为STR)
        int64 file_size = 18; // 文件/图片大小(字节)
        string video_url = 19; // 视频URL
        string audio_url = 21; // 语音URL
        int64 audio_time = 22; // 语音时长
        int64 sticker_item_id = 25; // 表情ID
        int64 sticker_pack_id = 26; // 表情包ID
        string call_text = 29; // 语音通话文字
        string call_status_text = 32; // 语音通话状态文字
        int64 width = 33; // 图片的宽度
        int64 height = 34; // 图片的高度
    }
    // 发送者信息
    message Sender {
        string chat_id = 1;
        int32 chat_type = 2;
        string name = 3;
        string avatar_url = 4;
        repeated string tag_old = 6;
        repeated Tag tag = 7;
    }
}
    
// 共用区结束

// 通过序列列出消息
message list_message_by_seq_send {
    int64 msg_seq = 3; // 从指定的msgSeq消息开始
    int64 chat_type = 4;
    string chat_id = 5;
}

message list_message_by_seq {
    Status status = 1;
    repeated Msg msg = 2;
    int32 total = 3; // 消息数
}

// 列出包含请求 msg_id 消息
message list_message_by_mid_seq_send {
    int64 msg_seq = 3; // 从指定的msgSeq开始
    int64 chat_type = 4;
    string chat_id = 5;
    int64 unknown = 6; // 不知道干啥的
    int64 msg_count = 7;
    string msg_id = 8;
}

message list_message_by_mid_seq {
    Status status = 1;
    repeated Msg msg = 2;
    int32 total = 3; // 消息数
}

// 列出消息
message list_message_send {
    int64 msg_count = 2; // 获取消息数
    string msg_id = 3; // 从指定消息ID开始
    int64 chat_type = 4; // 对象类型
    string chat_id = 5; // 对象ID
}

message list_message { // 其实可以和 list-message-by-seq共用的。
    Status status = 1;
    repeated Msg msg = 2;
}

// 发送消息
message send_message_send {
    string msg_id = 2; // 信息ID
    string chat_id = 3; // 欲发送到的信息对象
    int64 chat_type = 4; // 欲发送到的信息对象的类别，1-用户，2-群聊，3-机器人
    Content content = 5;
    message Content {
        string text = 1; // 信息文本
        string buttons = 2; // 按钮
        string file_name = 4; // 欲发送文件名称
        string file = 5; // 欲发送文件key/url
        repeated string mentioned_id = 6; // @用户ID
        string form = 7; // 表单消息
        string quote_msg_text = 8; // 引用信息文本
        string image = 9; // 欲发送图片key/url
        string msg_text1 = 11; // 信息文本
        string msg_text2 = 12; // 信息文本
        string expression_id = 15; // 个人收藏表情ID
        string temp_text2 = 16; // 未知
        string temp_text3 = 17; // 未知
        int64 file_size = 18; // 欲发送文件大小
        string video = 19; // 欲发送视频key/url(123.mp4)
        string audio = 21; // 语音key/url(123.m4a)
        int64 audio_time = 22; // 语音秒数
        string temp_text4 = 23; // 未知
        int64 temp_code1 = 24; // 未知
        int64 sticker_item_id = 25; // 表情ID
        int64 sticker_pack_id = 26; // 表情包ID
        string room_name = 29; // 语音房间发送显示信息的文本
    }
    int64 content_type = 6; // 信息类别，1-文本，2-图片，3-markdown，4-文件，5-表单，6-文章，7-表情，8-html，11-语音，13-语音通话
    int64 command_id = 7; // 所使用命令ID
    string quote_msg_id = 8; // 引用信息ID
    Media media = 9;
    message Media { // 在media发送对象为，图片/音频/视频
        string file_key = 1; // 发送对象key(就是上传后七牛对象存储给你返回的file_key)
        string file_hash = 2; // 发送对象上传返回哈希
        string file_type = 3; // 发送对象类别，image/jpeg-图片，video/mp4-音频
        int64 image_height = 5; // 图片高度
        int64 image_width = 6; // 图片宽度
        int64 file_size = 7; // 发送对象大小
        string file_key2 = 8; // 发送对象key,和1一样,据说不写会报错
        string file_suffix = 9; // 发送对象后缀名
    }
}

message send_message {
    Status status = 1;
}

// 撤回消息
message recall_msg_send {
    string msg_id = 2; // 信息ID
    string chat_id = 3; // 信息所属对象ID
    int64 chat_type = 4; // 信息所属对象类型, 1-用户 2-群聊 3-机器人
}

message recall_msg {
    Status status = 1;
}

// 批量撤回消息
message recall_msg_batch_send {
    repeated string msg_id = 2; // 信息ID
    string chat_id = 3; // 信息所属对象ID
    int64 chat_type = 4; // 信息所属对象类型, 1-用户 2-群聊 3-机器人
}

message recall_msg_batch {
    Status status = 1;
}

// 编辑消息
message edit_message_send {
    string msg_id = 2;
    string chat_id = 3;
    int32 chat_type = 4;
    Content content = 5;
    message Content {
        string text = 1;
        string buttons = 2;
        string quote_msg_text = 8;
    }
    int64 content_type = 6; // 信息类别，1-文本，3-markdown，8-html
    string quote_msg_id = 8; // 引用信息ID
}

message edit_message {
    Status status = 1;
}

// 通过按钮事件点击消息
message button_report_send {
    string msg_id = 2; // 信息ID
    int64 chat_type = 3; // 对象类型, 1-用户 2-群聊 3-机器人
    string chat_id = 4; // 对象ID
    string user_id = 5; // 按钮事件发送者ID
    string button_value = 6; // 欲点击按钮的值
}